Identity verification vendor integration guide (#5412)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/idv-integration/main/index.md

---
title: Integrate Okta with identity verification vendors
excerpt: How IDV vendors can integrate with Okta
layout: Guides
---

> Developers and organizations can use identity verification (IDV) vendors to integrate IDV flows into their authentication and authorization processes.

Reviewer: ```suggestion
Developers and orgs can use identity verification (IDV) vendors to integrate IDV flows into their authentication and authorization processes.
```

Reviewer: Organization works for me!

This guide shows developers of IDV vendors how their IDV solutions integrate with Okta orgs.

---

#### Learning outcome

> Understand the standardized process of how IDV vendors integrate with Okta

Reviewer: ```suggestion
Understand the standardized process of how IDV vendors integrate with Okta.
```

```json
 HTTP/1.1 400 Bad Request
 Content-Type: application/json
 Cache-Control: no-cache, no-store

 {
   "error": "invalid_request",
   "error_description":
     "The redirect_uri is not valid for the given client"
 }
```

#### Request parameters

| Parameter             | Description                                                                                                     | Param Type | DataType |
|-----------------------|-----------------------------------------------------------------------------------------------------------------|------------|----------|
| response_type         | Determines the type of flow. Okta uses `code` to identify the request as an `authorization_code` flow.          | String     | String   |
| client_id             | Unique identifier of the IDV vendor app. IDV vendors provide this value.                                         | String     | String   |
| client_secret         | A confidential key associated with the IDV vendor's `client_id`. IDV vendors provide this value.                | String     | String   |

> | code_challenge        | A cryptographic string derived from a code verifier. It’s used with the PKCE (Proof Key for Code Exchange) mechanism. | String     | String   |

Reviewer: ```suggestion
| code_challenge        | A cryptographic string derived from a code verifier. It’s used with the Proof Key for Code Exchange (PKCE) mechanism. | String     | String   |
```

| code_challenge_method | The transformation method that’s used to create the `code_challenge` from the code verifier. Okta uses the `S256` method. | String     | String   |
| scope                 | Defines the permissions or access levels that are requested.                                                   | String     | String   |

> | openid                | Identifies the request as an OpenID Connect request.                                                           | String     | String   |

Reviewer: ```suggestion
| openid                | Identifies the request as an OpenID Connect (OIDC) request.                                                           | String     | String   |
```

| profile               | Requests access to the end user's default profile claims.                                                      | String     | String   |
| identity_assurance    | Requests access to the `verified_claims` object.                                                               | String     | String   |
| idv_flow_{idv_flow_id} | Identifies a specific IDV flow from the IDV vendor. Replace `{idv_flow_id}` with the identifier of the flow. You can leave it out of the request if the `client_id` and `client_secret` are configured to represent a specific flow from the IDV vendor. | String     | String   |
| verified_claims       | Contains the `verification` and `claims` objects.                                                              | Object     | Object   |

> | verification          | Specifies the parameters and requirements for verifying the claims. Okta uses both the `trust_framework` and `assurance_level` properties. See the OpenID Connect definition of the [verification](https://openid.bitbucket.io/ekyc/openid-ida-verified-claims.html#name-verification-element) property. | Object     | Object   |

Reviewer: ```suggestion
| verification          | Specifies the parameters and requirements for verifying the claims. Okta uses both the `trust_framework` and `assurance_level` properties. See the OIDC definition of the [verification](https://openid.bitbucket.io/ekyc/openid-ida-verified-claims.html#name-verification-element) property. | Object     | Object   |
```

> | trust_framework       | Identifies the trust framework that provides assurance about the verified attributes. Okta sets `IDV_DELEGATED` as the default value. This value delegates identity verification and the assurance policy to the IDV vendor. The IDV vendor is then responsible for verifying user identities and sends the results back to Okta. `IDV_DELEGATED` is currently the only supported trust framework. See the OpenID Connect definition of the [trust_framework](https://openid.net/specs/openid-ida-verified-claims-1_0.html#section-5.4.2-5) property. | String     | String   |

Reviewer: ```suggestion
| trust_framework       | Identifies the trust framework that provides assurance about the verified attributes. Okta sets `IDV_DELEGATED` as the default value. This value delegates identity verification and the assurance policy to the IDV vendor. The IDV vendor is then responsible for verifying user identities and sends the results back to Okta. `IDV_DELEGATED` is currently the only supported trust framework. See the OIDC definition of the [trust_framework](https://openid.net/specs/openid-ida-verified-claims-1_0.html#section-5.4.2-5) property. | String     | String   |
```

> | assurance_level       | Identifies the assurance level that's required for the identity claims of the user. The IDV must map their verification results to the possible assurance levels, `VERIFIED` or `FAILED`. For a successful verification, the IDV vendor must pass  `VERIFIED` as the `assurance_level`. For a failed verification, the IDV vendor must pass `FAILED` or a `null` value. See the OpenID Connect definition of the [assurance_level](https://openid.net/specs/openid-ida-verified-claims-1_0.html#section-5.4.2-6.1.1) property. | String     | String   |

Reviewer: ```suggestion
| assurance_level       | Identifies the assurance level that's required for the identity claims of the user. The IDV must map their verification results to the possible assurance levels, `VERIFIED` or `FAILED`. For a successful verification, the IDV vendor must pass  `VERIFIED` as the `assurance_level`. For a failed verification, the IDV vendor must pass `FAILED` or a `null` value. See the OIDC definition of the [assurance_level](https://openid.net/specs/openid-ida-verified-claims-1_0.html#section-5.4.2-6.1.1) property. | String     | String   |
```

> | claims                | Contains user-specific attributes. `given_name` and `family_name` are currently the only supported claims. See the OpenID Connect definition of the [claims](https://openid.net/specs/openid-ida-verified-claims-1_0.html#name-claims-element) property. | Object     | Object   |

Reviewer: ```suggestion
| claims                | Contains user-specific attributes. `given_name` and `family_name` are currently the only supported claims. See the OIDC definition of the [claims](https://openid.net/specs/openid-ida-verified-claims-1_0.html#name-claims-element) property. | Object     | Object   |
```

### User's browser sends GET /oauth2/authorize request to IDV vendor

The user's browser sends the GET request to the IDV vendor's OAuth 2.0 authorization endpoint. The IDV vendor evaluates the `request_uri` so that the IDV flow can start.

> **Note:** For IDV vendors, Okta recommends renaming the authorization endpoint so that it's clear the endpoint is used for identity verification. For example, rename the endpoint to `/oauth2/idv-authorize`.

#### GET /oauth2/authorize request example

```json
GET /oauth2/idv-authorize?request_uri=urn:ietf:params:oauth:request_uri:6esc_11ACC5bwc014ltc14eY22c HTTP/1.1

Host: idv-vendor.com
```

#### GET /oauth2/authorize error response

The error response for an unsuccessful `GET /oauth2/authorize` request uses this [structure](https://www.rfc-editor.org/rfc/rfc6749#section-4.1.2.1).

```json
HTTP/1.1 302
GET /idp/identity-verification/callback?error=invalid_request&state=30pqcSFzH7H0bIftWwYRbNNwbpOpfY-W
```

### User completes the IDV flow

> The user is then redirected to the IDV vendor app and completes the IDV flow.

Reviewer: ```suggestion
The user is then redirected to the IDV vendor app to complete the IDV flow.
```

If either of the `trust_framework` or `assurance_level` attributes aren't included in the `id_token` [response](#idv-vendor-id-token-response-example) or if they have `null` or unexpected values, then Okta marks the policy evaluation as unsuccessful.

#### Failed policy evaluation example

For example, a user completes an IDV flow but provides a name that doesn't match the `given_name` value. The `assurance_level` value is sent back as `FAILED` in the `id_token` response. And the `given_name` value is returned as `null`.

```json
HTTP/1.1 200 OK
Content-Type: application/json
Cache-Control: no-cache, no-store

{
  "id_token": {
    "verified_claims": [
      "verification": {
        "trust_framework": "IDV-DELEGATED",
        "assurance_level": "FAILED",
        "time": "2024-10-07T10:00:00Z"
      },
      "claims": {
        "given_name": null,
        "family_name": "Jones"
      }
    ]
  }
}
```

## Identity verification events

After Okta completes the policy evaluation, regardless of the result, an event is triggered, `user.identity_verification`. Okta admins can view this event in the System Log. See [Event Types](/docs/reference/api/event-type).

> There are two possible results for the event: `ALLOW` (user is successfully verified) or `DENY` (user isn’t verified).

Reviewer: ```suggestion
There are two possible results for the event: `ALLOW` (user is successfully verified) or `DENY` (user isn't verified).
```
